india_stocks_api/brokers/angel.py

```
@@ -57,31 +53,47 @@ def authenticate(self) -> bool:
         Login using SmartAPI.
         Updates the internal context with the session token.
+        Also persists credentials to creds.json **only** after successful auth.
         """
         # Hack: The ported code reads BROKER_API_KEY from os.environ
         # We must set it here for the internal function to work.
         # Ideally, we would patch the internal code to use context.get_api_key()
-        os.environ['BROKER_API_KEY'] = self.api_key
+        if not all([self.api_key, self.client_code, self.password, self.totp_key]):
+            creds = context.get_credentials(self.broker_name)
+            if not creds:
+                raise RuntimeError("No credentials provided or stored.")
```

Ruff (0.14.14)

Avoid specifying long messages outside the exception class

(TRY003)